### ISSUE TITLE
Client :timeout

### ISSUE THREAD

**Issue by philip-doctor** [http-kit/http-kit#327]
I was just trying to make sure I understood the :timeout parameter, I was reading here:
https://github.com/http-kit/http-kit/blob/2.2.0/src/java/org/httpkit/client/HttpClient.java#L159
and here
https://github.com/http-kit/http-kit/blob/2.2.0/src/java/org/httpkit/client/HttpClient.java#L344

It looks like "timeout is reset based on the last time we connected or read from the destination."

However I'm not clear, if I have a large body I'm trying to send via the client (and/or a particularly slow network connection) I don't see onProgress getting called for writes.  If my write takes longer than timeout it appears to me like I will also timeout. 

Is that a correct reading of the code?

Summary:
I will timeout if
1) I haven't connected within the specified timeout span
2) I haven't read within the specified timeout span
3) My body takes longer than the specified timeout span to write

Thanks for your time.

**Comment by ptaoussanis** [comment]
> Is that a correct reading of the code?

Yes, I believe that's all correct as per the current implementation.

Semantically, the timeout _appears_ to be intended to function as an inactivity timeout. I.e. you set a timeout of 10s to mean: trigger if there's inactivity of >= 10s.

> I don't see onProgress getting called for writes

But this part does appears to violate the previous intention. I.e. my expectation would be that write activity resets the timeout (as read activity does).


Haven't given this much though, so very open to input on this. Would it perhaps be more sensible to consistently treat all timeouts semantically as: trigger if the job's not done after this much time? I.e. no extensions on reads?

**Comment by ptaoussanis** [comment]
Have added a [TODO](https://github.com/http-kit/http-kit/blob/bdbb2b1134350064c46376bb02ef97a94b4c47b6/src/java/org/httpkit/client/HttpClient.java#L190) in the interests of consistency.

PR welcome if anyone felt like tackling this!

**Comment by philip-doctor** [comment]
Thanks for the response, I agree with

> But this part does appears to violate the previous intention.

Having write update onProgress would feel ideologically consistent (and tangentially benefit my use case). 

Additionally having write update onProgress is providing me with functionality that's a lot harder for me to reimplemented within my application.  By comparison, I can trivially timeout an entire call within my application I desired to have a hard limit of X seconds for the entire round trip no matter what.

So I guess that's saying, it seems consistent to me to have write update onProgress and additionally provides me with value I have a harder time reimplementing myself within my application. I would be +1 for my part.
